File: src/test/java/com/pschlup/ta/indicators/SimpleMovingAverageTest.java
package com.pschlup.ta.indicators;

import com.pschlup.ta.helpers.MockTimeSeries;
import com.pschlup.ta.timeseries.UnstablePeriodException;
import org.junit.jupiter.api.Test;

import static com.google.common.truth.Truth.assertThat;
import static org.junit.jupiter.api.Assertions.*;

class SimpleMovingAverageTest {

  private double[] PRICES = new double[] {
      1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11,
  };

  @Test
  void shouldComputeAverage() {
    SimpleMovingAverage sma = SimpleMovingAverage.of(MockTimeSeries.of(PRICES).closePrice()).withLength(10);
    assertThat(sma.valueAt(1)).isEqualTo(5.5);
    assertThat(sma.valueAt(0)).isEqualTo(6.5);
  }

  @Test
  void shouldThrowIfUnstable() {
    SimpleMovingAverage sma = SimpleMovingAverage.of(MockTimeSeries.of(PRICES).closePrice()).withLength(20);
    assertThrows(UnstablePeriodException.class, () -> sma.valueAt(0));
  }
}

File: src/test/kotlin/com/pschlup/ta/indicators/SimpleMovingAverageTest.kt
package com.pschlup.ta.indicators

import com.google.common.truth.Truth.assertThat
import com.pschlup.ta.helpers.FakeTimeSeries
import com.pschlup.ta.timeseries.UnstablePeriodException
import org.junit.jupiter.api.Assertions.assertThrows
import org.junit.jupiter.api.Test

internal class SimpleMovingAverageTest {
  private val timeSeries = FakeTimeSeries.of(
    doubleArrayOf(1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0, 10.0, 11.0)
  )

  @Test
  fun shouldComputeAverage() {
    val sma = timeSeries.closePrice.sma(10)

    assertThat(sma.valueAt(1)).isEqualTo(5.5)
    assertThat(sma.valueAt(0)).isEqualTo(6.5)
  }

  @Test
  fun shouldThrowIfUnstable() {
    val sma = timeSeries.closePrice.sma(20)

    assertThrows(UnstablePeriodException::class.java) { sma.valueAt(0) }
  }
}
